InterGalacticCurrencyConvertor/src/main/java/numerals/RomanNumerals.java

```diff
@@ -1,145 +1,154 @@
-				} else if (inRNums.get(i-1).equals("v") || inRNums.get(i-1).equals("l") ||
-				           inRNums.get(i-1).equals("d")){
-					return vResult;
-				}
-				//System.out.println(inRNums.get(i-1) + " rise " + inRNums.get(i));
-			}
-			if (vRiseOcc > -1 && vRiseOcc < i && vMinimum < getDecimalVal(inRNums.get(i))) {
-				return vResult;
-			}
-			if (inRNums.get(i) != vCurRepRN) {
-				vCurRepRN = inRNums.get(i);
-				vCurRepRNCnt = 1;
-			} else {
-				++vCurRepRNCnt;
-			}
-			//System.out.println("middle");
-			if (vCurRepRNCnt > 1) {
-				if (!(inRNums.get(i).equals("i") || inRNums.get(i).equals("x") ||
-				    inRNums.get(i).equals("c") || inRNums.get(i).equals("m"))) {
-					return vResult;
-				} else if (vCurRepRNCnt == 4 && inRNums.get(i).equals(vCurRepRN)) {
-					return vResult;
-				}
-			} else if (vOccurrenceMap.get("i") > 4 || vOccurrenceMap.get("x") > 4 ||
-			           vOccurrenceMap.get("c") > 4 || vOccurrenceMap.get("m") > 4 ||
-			           vOccurrenceMap.get("v") > 1 || vOccurrenceMap.get("l") > 1 ||
-			           vOccurrenceMap.get("d") > 1) {
-				return vResult;
-			}
-
-			vBefore = vAfter;
-			if (vMinimum > getDecimalVal(inRNums.get(i))) {
-				vMinimum = getDecimalVal(inRNums.get(i));
-			}
-			//System.out.println("end " + vCurRepRNCnt);
-		}
-		vResult = true;
-		return vResult;
+	private boolean subtractionRuleFail(Character curChar,
+	                                    Character prevChar,
+	                                    int idx,
+	                                    List<Integer> rSymbolValues) {
+	  boolean isInvalid = true;
+	  int curVal = toDecimal(curChar);
+	  if (String.valueOf(prevChar).matches("[vld]")) {
+      return isInvalid;
+    } else if (prevChar.equals('i') && String.valueOf(curChar).matches("[^vx]")) {
+      return isInvalid;
+    } else if (prevChar.equals('x') && String.valueOf(curChar).matches("[^lc]")) {
+      return isInvalid;
+    } else if (prevChar.equals('c') && String.valueOf(curChar).matches("[^dm]")) {
+      return isInvalid;
+    }
+    for (int j = idx - 2; j >= 0; --j) {
+      if (rSymbolValues.get(j) < curVal) {
+        return isInvalid;
+      }
+    }
+    return !isInvalid;
 	}
-	public boolean isRomanNumeral(String inRNum) {
-		if (inRNum == null) {
-			return false;
-		}
-		return mMap.containsKey(inRNum);
+  private boolean additionRuleFail(Character curChar,
+                                   int idx,
+                                   List<Integer> values) {
+    boolean isInvalid = true;
+    int curVal = toDecimal(curChar);
+    for (int j = idx - 2; j >= 0; --j) {
+      if (values.get(j) <= curVal) {
+        return isInvalid;
+      }
+    }
+    return !isInvalid;
+  }
+  private boolean repetitionRuleFail(Character curChar,
+                                     int idx,
+                                     List<Integer> values,
+                                     HashMap<Character, Integer> charCntMap) {
+    boolean isInvalid = true;
```

Comment:
Same comment of changing pattern of 'isInvalid' here.

Reply:
done